sdk/schemaregistry/azure-schemaregistry-avroencoder/azure/schemaregistry/encoder/avroencoder/_schema_registry_avro_encoder.py

```diff
@@ -26,25 +26,40 @@
 import logging
 from functools import lru_cache
 from io import BytesIO
-from typing import Any, Dict, Mapping, Optional, Union, Callable
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Mapping,
+    Optional,
+    Type,
+    overload,
+    TypeVar,
+)
+from avro.errors import SchemaResolutionException
 
-from .exceptions import (  # pylint: disable=import-error
-    SchemaParseError,
-    SchemaEncodeError,
-    SchemaDecodeError,
+from ._exceptions import (  # pylint: disable=import-error
+    InvalidContentError,
+    InvalidSchemaError
+)
+from ._apache_avro_encoder import (  # pylint: disable=import-error
+    ApacheAvroObjectEncoder as AvroObjectEncoder,
 )
-from ._apache_avro_encoder import ApacheAvroObjectEncoder as AvroObjectEncoder  # pylint: disable=import-error
-from ._message_protocol import MessageContent, MessageType  # pylint: disable=import-error
-from ._constants import (   # pylint: disable=import-error
-    SCHEMA_ID_START_INDEX,
-    SCHEMA_ID_LENGTH,
-    CONTENT_START_INDEX,
+from ._message_protocol import (  # pylint: disable=import-error
+    MessageContent,
+    MessageType,
+)
+from ._constants import (  # pylint: disable=import-error
     AVRO_MIME_TYPE,
-    RECORD_FORMAT_IDENTIFIER_LENGTH,
 )
 
+if TYPE_CHECKING:
+    from azure.schemaregistry import SchemaRegistryClient
+
 _LOGGER = logging.getLogger(__name__)
 
+T = TypeVar("T", bound=MessageType) # pylint: disable=invalid-name
```

Comment:
I feel we could have a better descriptive name than a single character "T"..

Reply:
thoughts on names: MessageT?

Reply:
I'm poor at naming..
was wondering if this name kept private to our users? then we could name whatever we want with it.. I could live with MessageTypeT

Reply:
I'm thinking that it is, but generating the doc rn to check.

Reply:
looks like this typevar doesn't show in the docs, so we're good naming it anything I think!